### DISCUSSION TITLE
Using the package to convert image formats

### DISCUSSION THREAD

**Discussion by puntiz** [hm21/pro_image_editor#244] [Q&A] [answered]
Hi,

I realize that this is not really the primary use case for this package, however, I was wondering since you use the _Image_ package and you have a fantastic web workers implementation that makes the Image library usable on web without just giant UI freezes while it tries to detect file formats etc, is there a way for me to leverage just that aspect of your library. I have a  Tiff image and I run it through you library and get a JPG or PNG in return skipping the UI entirely ?

**Comment by hm21** [comment, marked as answer]
Am I understanding correctly that you only want to change the image format without using the image editor? 
If so, as you mentioned, this isn’t the primary use of the package, so there’s no straightforward way to achieve it. However, there is an exported function typically used for thumbnail generation that allows users to later generate a high-resolution image. You could also add your image with the desired output format, which would convert it in a web worker, handling just simple calculations in the main thread.


```dart
import 'dart:async';
import 'dart:typed_data';
import 'dart:ui' as ui;
import 'package:flutter/widgets.dart';
import 'package:pro_image_editor/pro_image_editor.dart';

Future<Uint8List?> convert({
  required BuildContext context,
  required Uint8List rawImage,
  required OutputFormat format,
}) async {
  Future<ui.Image> convertImage(Uint8List bytes) async {
    final Completer<ui.Image> completer = Completer();
    ui.decodeImageFromList(bytes, (ui.Image img) {
      completer.complete(img);
    });
    return completer.future;
  }

  var convertedImage = await convertImage(rawImage);

  if (!context.mounted) return null;

  return generateHighQualityImage(
    convertedImage,
    context: context,
    configs: ImageGenerationConfigs(
      outputFormat: format,
    ),
  );
}
```

**Reply by puntiz** [reply]
Hey @hm21, wanted to check with you again on this specific requirement. Last I tried the methods kept returning null value. 6 months on is this still the way to achieve the stated requirement ? Thanks again.

**Reply by hm21** [reply]
Yes, that's still the only way to achieve this. Since you mentioned it's just returning null, could you share a quick code example of how you're currently handling it? I'd like to take a look and see where the issue might be.

**Reply by puntiz** [reply]
Have the code snippet you provided added as is and to call the function I use the below code. I have validated that convertedImage is generated and generateHighQualityImage is called but "Encoded" from below code is never printed to console so the '.then' never gets called i.e., future never completes.

`          convert(rawImage: imageToDownload, format: OutputFormat.jpg)
              .then((imageJpg) {
            print("Encoded");
            //Remaining logic here
          });`

**Reply by hm21** [reply]
To simplify things, I introduced a new singleton class called `ImageConverter` in version `9.7.0`. This makes it easier to convert image formats, as shown in the example below:

```dart
var result = await ImageConverter.instance.convertFormat(
  image: EditorImage(byteArray: bytes),
  format: OutputFormat.jpg,
);
```

**Reply by puntiz** [reply]
Works perfectly .. Thank you for accommodating the request.
